### PR TITLE
Add `/dev/disk/by-id` link tests

### PR DESCRIPTION
**Pull request by hamistao**
This adds tests for checking if `/dev/disk/by-id` links are being properly set up for different kinds of block devices. Also moves non pool specific tests away from `storage-vm`.
This is needed for [#13672](https://github.com/canonical/lxd/pull/13672) and [#13671](https://github.com/canonical/lxd/pull/13671).

### CODE REVIEW

File: tests/storage-vm
Comment (simondeziel) [inline tests/storage-vm:306]: IIRC, we want to use `/dev/disk/by-id` to avoid depending on automatically allocated `/dev/sdX` style as the later are not guarantied to be stable.

File: tests/storage-vm
Comment (simondeziel) [inline tests/storage-vm:305]:
```suggestion
        [ "$(lxc exec v1 -- stat -c "%F" /dev/disk/by-id/scsi-0QEMU_QEMU_HARDDISK_lxd_block1ro)" = "symbolic link" ]
```

This is faster and easier to read IMHO in logs.

File: tests/storage-vm
Comment (hamistao) [reply tests/storage-vm]: I figured guaranteeing the link in properly set up took care of the problem, would it be better to just substitute every `/dev/sd*` reference?

File: tests/storage-volumes-vm
Comment (tomponline) [inline tests/storage-volumes-vm:132]: These lines have changed multiple times in the same PR - please dont do this.

Each commit in this PR should be the final version of that logical change.

Additionally these line changes do not relate to the commit message:

>  tests/storage-volumes-vm: Check links to /dev/s*

**Review by tomponline — CHANGES_REQUESTED** [review]
There seems to be some commits that are iterating on changes of the same lines - please dont do this. As discussed before each line change should be the final version of a logical change - not an iteration.

Secondly, we agreed on Friday that we would move the "Checking additional disk device support" tests from storage-vm to storage-disks-vm so as to keep the external disks together in one file and so as to avoid repeating them for each pool driver unnecessarily, however I do not see this change.

**Comment by tomponline** [comment]
@hamistao additionally please can you also add some tests to the storage-disks-vm file that checks for devices passed with long & escaped names that are longer than what udev uses, so we can check the udev /dev/disk/by-id path is unchanged. Ta

**Comment by hamistao** [comment]
@tomponline I am planning to include the discussed moving of the tests on a later PR along with other similar changes to make this simpler.
Also, using long names don't serve a lot of purpose here because the name is trimmed to about 16 characters when added to the `/dev/disk/by-id` path, if you think it still makes sense to use them I can make the changes. As for escaped names, I believe I am using one.

File: tests/storage-volumes-vm
Comment (hamistao) [reply tests/storage-volumes-vm:132]: Sorry, this one was an accident during the last rebase, fixed

**Comment by hamistao** [comment]
@tomponline The new long name tests were added, feel free to take a look at this when you can. I believe this is failing the tests due to unrelated problems with `latest/edge`.

**Comment by tomponline** [comment]
Thanks, unfortunately as latest/edge is now broken ill hold off on this until @mihalicyn and I can unblock it.

**Comment by tomponline** [comment]
> The new long name tests were added

You mentioned some devices didn't work with long names - please can you elaborate on this.

**Comment by hamistao** [comment]
@tomponline It seemed like using long names could break the devices but in fact, the paths on `/dev/` and `/dev/disk/by-id` behave differently. For example, on `ubunt-minimal` VMs adding two devices with names that share a 16 character prefix, the `/dev/disk/by-id` path of the first would be overwritten by the second, leaving the first without a link on `/dev/disk/by-id`.
For the purpose of these tests, thay are not affected by this behavior.

**Comment by hamistao** [comment]
> Thanks, unfortunately as latest/edge is now broken ill hold off on this until @mihalicyn and I can unblock it.
No problem!

File: tests/storage-disks-vm
Comment (tomponline) [inline tests/storage-disks-vm:161]: Why do we need to do this (your commit message didnt explain the why, only the what)?

**Comment by hamistao** [comment]
@tomponline I will push new changes removing the tests that use `readlink` to check links to `/dev/sd*`, as was discussed earlier with @simondeziel

File: tests/storage-disks-vm
Comment (tomponline) [reply tests/storage-disks-vm:161]: Why did the previous form of the tests not require this stuff to be turned off?

**Comment by tomponline** [comment]
Good news, the latest/edge is unblocked and the tests are green.

Just need to understand about the restrictions being relaxed and then we can merge.

File: tests/storage-disks-vm
Comment (hamistao) [reply tests/storage-disks-vm:161]: this became a requirement after adding new long name tests in the previous commit (this time with a block device that uses `loopdev`). Please feel free to suggest alternative ways to perform this tests if you would like.

Comment (hamistao) [reply tests/storage-disks-vm:161]: yes, I just figured it made sense to keep them together as they are unrelated to the restriction tests

File: tests/storage-disks-vm
Comment (simondeziel) [inline tests/storage-disks-vm:188]: This is still hardcoding `/dev/sdX` which is what we should avoid. I'd re-order the test to have a `readlink` call to know which `/dev/sdX` it points to and then use that variable later on.

**Review by simondeziel — DISMISSED** [review]
Please remove remaining traces of `/dev/sdX` everywhere.

File: tests/storage-disks-vm
Comment (hamistao) [reply tests/storage-disks-vm:188]: I was just making this exact change :). Will be pushing it shortly, thanks!

Comment (hamistao) [reply tests/storage-disks-vm:188]: I was just making this exact change :). Will be pushing it shortly, thanks!

File: tests/storage-vm
Comment (simondeziel) [inline tests/storage-vm:347]: I don't think we need to do the symlink deference manually, do we?

```suggestion
        lxc exec v1 -- test -b /dev/disk/by-id/scsi-0QEMU_QEMU_HARDDISK_lxd_block1ro
```

Comment (hamistao) [reply tests/storage-vm]: fixed, thanks for the insight!

File: tests/storage-vm
Comment (tomponline) [inline tests/storage-vm:382]: Messy spacing

File: tests/storage-disks-vm
Comment (tomponline) [inline tests/storage-disks-vm:179]: I would prefer to keep these tests in a place where the project restrictions were still enabled so we are testing that this works with project restrictions (as we are now).

File: tests/storage-disks-vm
Comment (tomponline) [inline tests/storage-disks-vm]: What was the reason for the / change here?

Comment (hamistao) [reply tests/storage-disks-vm]: The reason was to make `/` show on the `/dev/disk/by-id` path so we can check for escaping

**Review by tomponline — APPROVED** [review]
Thanks!

**Pull request merged**
